File: Stack/Python/Design a Stack With Increment Operation.py
# Problem link: https://leetcode.com/problems/design-a-stack-with-increment-operation/description
# Time: O(1)
# Space: O(n)
class CustomStack:

    def __init__(self, maxSize: int):
        self.stack = [0] * maxSize
        self.inc = [0] * maxSize
        self.top = -1

    def push(self, x: int) -> None:
        if len(self.stack) > self.top + 1:
            self.top += 1
            self.stack[self.top] = x

    def pop(self) -> int:
        if self.top >= 0:
            value = self.stack[self.top] + self.inc[self.top]
            if self.top > 0:
                self.inc[self.top - 1] += self.inc[self.top]
            self.inc[self.top] = 0
            self.top -= 1
            return value
        return -1

    def increment(self, k: int, val: int) -> None:
        if self.top >= 0:
            index = min(k - 1, self.top)
            self.inc[index] += val
        

File: Stack/Java/Design a Stack With Increment Operation.java
// Problem link: https://leetcode.com/problems/design-a-stack-with-increment-operation/description
// Time: O(1)
// Space: O(n)
class CustomStack {

    int[] stack, inc;
    int top;

    public CustomStack(int maxSize) {
        this.stack = new int[maxSize];
        this.inc = new int[maxSize];
        this.top = -1;
    }
    
    public void push(int x) {
        if(this.stack.length - 1 > this.top) {
            this.stack[++this.top] = x;
        }
    }
    
    public int pop() {
        if(this.top >= 0) {
            int value = this.stack[this.top] + this.inc[this.top];

            if(this.top > 0) {
                this.inc[this.top - 1] += this.inc[this.top];
            }

            this.inc[this.top--] = 0;
            return value;
        }
        return -1;
        
    }
    
    public void increment(int k, int val) {
        if(this.top >= 0) {
            this.inc[Math.min(this.top, k - 1)] += val;
        }
        
    }
}
